File: trees/3.js
let arr = [];

function insert(data) {
  arr.push(data);
  let pos = arr.length-1;
  heapifyUp(pos)
}

function heapifyUp(pos) {
  if (pos > -1) {
  let parent = Math.floor((pos + 1) / 2) - 1;
  if (arr[pos] < arr[parent]) {
    arr[parent] =  arr[parent] +  arr[pos];
    arr[pos] = arr[parent] -  arr[pos];
    arr[parent] = arr[parent] -  arr[pos];
  }
  heapifyUp(parent);
}
return;
}

function heapifyDown(pos) {
  let left = 2* pos + 1;
  let right = 2* pos + 2;
  let parent = pos;
  if (arr[left] && arr[left] < arr[parent]) {
    arr[parent] =  arr[parent] +  arr[left];
    arr[left] = arr[parent] -  arr[left];
    arr[parent] = arr[parent] -  arr[left];
  }
  if (arr[right]  && arr[right] < arr[parent]) {
    arr[parent] =  arr[parent] +  arr[right];
    arr[right] = arr[parent] -  arr[right];
    arr[parent] = arr[parent] -  arr[right];
  }
  if (left > arr.length-1 || right > arr.length-1) {
    return;
  } else {
    heapifyDown(left);
    heapifyDown(right);
  }
}

function remove(el) {
  for (var i = 0; i < arr.length; i++) {
   if (el == arr[i]) {
     let end = arr.pop();
     arr[i]  = end;
     let parent = Math.floor((i + 1) / 2) - 1;
     if (arr[i] < arr[parent]) {
      heapifyUp(i);
     } else {
      heapifyDown(i);
     }  
     break;
   }
}}

function isEmpty() {
  return arr.length == -1;
}

insert(5);
insert(4);
insert(8);
insert(6);
insert(1);
insert(14);
insert(2);
insert(7);
console.log("final min heap", arr);
remove(6);
console.log("final min heap after remove 6", arr);
console.log("is Heap Empty", isEmpty());

File: trees/3.go
package main

import (
	"fmt"
	"math"
)

var arr []int

func insertData(array []int) {
	for i := 0; i < len(array); i++ {
		// fmt.Println("parent", array)
		heapifyUp(array[i])
	}
}

func heapifyUp(pos int) {
	if pos > -1 {

		parent := int(math.Floor(float64((pos+1)/2)) - 1)

		fmt.Println("parent", pos, parent)
		if parent > -1 && arr[pos] < arr[parent] {
			arr[parent] = arr[parent] + arr[pos]
			arr[pos] = arr[parent] - arr[pos]
			arr[parent] = arr[parent] - arr[pos]
		}
		fmt.Println("arr", arr)
		heapifyUp(parent)
	}
	return
}

// func heapifydown(pos int) {

// 	fmt.Println("pos", pos)

// }

func main() {
	input := []int{11, 12, 13, 41, 5, 16, 17, 81, 19, 110}
	// fmt.Println("input", insertData)
	insertData(input)
	fmt.Println("out ->", arr)
	// heapifydown(input)
}
